### PR TITLE
module3 task done

### PR DESCRIPTION
**Pull request by never47**
About the tests: I don’t check successes and failures, all the tests are of the same type, because all the main logic lies in the repository classes (I will check this in the next module).

### CODE REVIEW

File: .idea/.gitignore
Comment (HannaVatlina) [inline .idea/.gitignore:1]: Please do not store system folder from IntelliJ Idea, add this to .gitignore
*.xml

File: src/main/java/springTask/Utils.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+
+import java.util.List;
+import java.util.Random;
+
+public class Utils {
+    private final static String ALLOWED_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    private final static int PASSWORD_LENGTH = 10;
+
+    public static int countUsernameSerialNumber(List<String> usernames, String firstName, String lastName){
```
Comment (HannaVatlina) [inline src/main/java/springTask/Utils.java:11]: what this method do?

File: src/main/java/springTask/Utils.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package springTask;
+
+
+import java.util.List;
+import java.util.Random;
+
+public class Utils {
+    private final static String ALLOWED_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
```
Comment (HannaVatlina) [inline src/main/java/springTask/Utils.java:8]: A strong password is a password that contains not only letters, but also numbers and special characters.

File: src/main/java/springTask/dao/StorageMemory.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class StorageMemory {
+    private TraineeRepository traineeRepository;
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/StorageMemory.java:11]: It's interesting, why you need repository in your application?

File: src/main/java/springTask/dao/TraineeRepository.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    public Set<String> getUsernames(){
+        return dataBase.keySet();
+    }
+
+    public void create(Trainee trainee){
+        try{
+            if(trainee.getUsername()==null){
+                throw new IllegalArgumentException("username cant be equal to null");
+            }
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TraineeRepository.java:65]: you can move this check on the service layer

File: src/main/java/springTask/dao/TraineeRepository.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            dataBase.put(trainee.getUsername(), trainee);
+        }catch (IllegalArgumentException e){
+            logger.warn(e);
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    public Trainee select(String key){
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TraineeRepository.java:73]: better to name like getByUsername or selectByUsername

File: src/main/java/springTask/dao/TraineeRepository.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            return null;
+        }
+        return dataBase.get(key);
+    }
+
+    public void delete(String key){
+        try{
+            if(!dataBase.containsKey(key)){
+                throw new IllegalArgumentException("no trainee with such key");
+            }
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TraineeRepository.java:84]: the same, you can move this validation on service layer

File: src/main/java/springTask/dao/TraineeRepository.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            dataBase.remove(key);
+        }catch (IllegalArgumentException e){
+            logger.warn(e);
+            throw new IllegalArgumentException(e);
+        }
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TraineeRepository.java:89]: do you really need to catch the exception you raised a couple of lines ago?

File: src/main/java/springTask/dao/TraineeRepository.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                throw new IllegalArgumentException("username cant be equal to null");
+            }
+            if(!dataBase.containsKey(key)){
+                throw new IllegalArgumentException("no trainee with such key");
+            }
+            dataBase.put(key, trainee);
+        }catch (IllegalArgumentException e){
+            logger.warn(e);
+            throw new IllegalArgumentException(e);
+        }
```
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TraineeRepository.java:104]: the same about validation and catch exception

File: src/main/java/springTask/dao/TrainerRepository.java
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TrainerRepository.java:1]: same comments as for TraineeRepository

File: src/main/java/springTask/dao/TrainingRepository.java
Comment (HannaVatlina) [inline src/main/java/springTask/dao/TrainingRepository.java:1]: same comments as for TraineeRepository

File: src/main/java/springTask/entity/TrainingType.java
Comment (HannaVatlina) [inline src/main/java/springTask/entity/TrainingType.java:1]: Later you will need to storage this in DB

File: src/main/java/springTask/service/ServiceFacade.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package springTask.service;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ServiceFacade {
```
Comment (HannaVatlina) [inline src/main/java/springTask/service/ServiceFacade.java:7]: What does this component do?

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -0,0 +1,3 @@
+traineeDB.csv:src/main/resources/traineeDB.csv
```
Comment (HannaVatlina) [inline src/main/resources/application.properties:1]: in property files is used equal
`traineeDB.csv=src/main/resources/traineeDB.csv`

File: src/test/java/springTask/service/TraineeServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    void selectTrainee() {
+        Trainee traineeExpected = new Trainee("Jack", "Brown", "jack.brown", "pass123"
+                ,true, java.time.LocalDate.of(1999, 1 ,1),"California", "jack.brown");
+
+        when(storageMemory.getTraineeRepository().select("jack.brown")).thenReturn(traineeExpected);
+
+        Trainee traineeActual = traineeService.selectTrainee("jack.brown");
+
+        verify(traineeRepository).select("jack.brown");
```
Comment (HannaVatlina) [inline src/test/java/springTask/service/TraineeServiceTest.java:56]: Here you can add additional checks for the received traineeActual object

File: src/test/java/springTask/service/TraineeServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        Trainee traineeExpected = new Trainee("Jack", "Brown", true,
+                java.time.LocalDate.of(1999, 1, 1), "California");
+
+        List<String> usernames = List.of("magamed.guseinovi");
+
+        when(storageMemory.getAllUsernames()).thenReturn(usernames);
+
+        traineeService.createTrainee(traineeExpected);
+
+        verify(traineeRepository).create(traineeExpected);
```
Comment (HannaVatlina) [inline src/test/java/springTask/service/TraineeServiceTest.java:44]: Here you can add checks that the object was created

File: src/main/java/springTask/App.java
Comment (HannaVatlina) [inline src/main/java/springTask/App.java:1]: Please add the ability to work with objects from the console
Scanner scanner = new Scanner(System.in);

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-context</artifactId>
+      <version>6.1.11</version>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <version>5.11.0</version>
```
[the remark this answers is not in the document]
Comment (never47) [reply pom.xml:30]: done

File: src/main/java/springTask/Utils.java (hunk 1 above)
Comment (never47) [reply src/main/java/springTask/Utils.java:11]: This method is used to count next serial number for new nickName.

if there exists hanna.vatlina username in database, next one must be hanna.vatlina1, method counts that number and returns it

File: src/main/java/springTask/Utils.java (hunk 2 above)
Comment (never47) [reply src/main/java/springTask/Utils.java:8]: done, changed the function

File: src/main/java/springTask/dao/TraineeRepository.java (hunk 4 above)
Comment (never47) [reply src/main/java/springTask/dao/TraineeRepository.java:65]: done

File: src/main/java/springTask/dao/TraineeRepository.java (hunk 5 above)
Comment (never47) [reply src/main/java/springTask/dao/TraineeRepository.java:73]: done

**Pull request closed without merging**